### ISSUE TITLE
Invalid get index when updating batches via threads

### ISSUE THREAD

**Issue by griessbreilp** [Portponky/better-terrain#66]
Am not quite sure if this is an actual problem with Better Terrain or with my code!
I really enjoy Better Terrain, it already helped me quite a lot and is imo way better than the default that Godot has. Thank you really much for the plugin! 

![image](https://github.com/Portponky/better-terrain/assets/11736148/1d3ab22c-3b40-4b99-a362-62f01d713934)
So what happens is I create a noise filter, loop through two FORs for X and Y coordinates, calculate if they are too far away from my radius, and if not, check which noise I have on this Coord, and set a tile accordingly.

That leads to me being able to create circular (or which ever shape I want) islands with different terrains.
To make it more performant I only update the layers when the whole island is calculated.
I have overall ~6.400 - 14.400 Tiles I want to set and after that update.

It looks a bit like that:
- Start a thread to calculate the island size and which tiles to place
- set the tiles via better_terrain.set_cells on the main thread
- add the tiles that just were set to the update queue as a batch
- update the tile batch on a new thread

Because I use the same array and layer_id for update_terrain_cells as I do for set_cells, I am pretty sure that the tile exists.
![image](https://github.com/Portponky/better-terrain/assets/11736148/66671599-908e-4afe-a66d-f71fc537b4c2)

When I open up **a thread for each layer at once** with **"update_surrounding_cells" = false OR true (both cases)** it **fails**:
![image](https://github.com/Portponky/better-terrain/assets/11736148/fd3a44f2-b169-49f0-81ea-e3c42335f228)
It is btw not the same layer. Sometimes it is the third, sometimes the seventh, ... that doesn't seem to be the problem.

When I open up **a thread for each layer after each other** with **"updating_surrounding_cells" = false** it **fails**:
![image](https://github.com/Portponky/better-terrain/assets/11736148/155b5333-faff-44c2-a6e4-26e70b9adf85)

When I open up **a thread for each layer after each other** with **"updating_surrounding_cells" = true** it **succeeds**:
![image](https://github.com/Portponky/better-terrain/assets/11736148/274537e9-c227-4314-98c9-95ebf80f825f)

How I start the threads:
![image](https://github.com/Portponky/better-terrain/assets/11736148/c41fe971-9e64-44ac-8664-24702799e521)

How I kill the threads:
![image](https://github.com/Portponky/better-terrain/assets/11736148/9331078e-c47d-4b2f-ad7e-1e9845f7fec0)

I just change 
`if !tile_batches_to_update.is_empty() && !tile_batches_updating:`
to
`if !tile_batches_to_update.is_empty(): `
to start multiple threads at once

It could totally be that my thread work is crap and that this causes the error, but I don't think that it should crash at all, even if there is a problem.

**Comment by Portponky** [comment]
The Godot Tilemap has serious problems when updating tiles outside of the main thread. Sometimes it works, sometimes it doesn't. in Better Terrain, the `update_terrain_cells` function calls `Tilemap.set_cell`, so it also can't be used outside of the  main thread.

This isn't a huge problem because Better Terrain offers a threading system that will calculate the terrain in a thread pool, so it can be applied relatively quickly on the main thread.

Check out `create_terrain_changeset`, `is_terrain_changeset_ready` and `apply_terrain_changeset` functions.

**Comment by griessbreilp** [comment]
Mate, you are just nuts! Really appreciate it :D 
Thanks for the fast reply! 

So if I understand this correctly, I still have to set the tiles with BetterTerrain.set_cells or what ever function on the Main-Thread.
When the tiles are set I can use BetterTerrain.create_terrain_changeset with the Tilemap, the Layer ID and a Dictionary containing Coords to Tile ID.
Is the Tile ID correct? Or what is "terrain types"?

This returns a Dictionary containing all the infos and the WorkerThreadPool Group Task. Which, if I might say, is just genious.
And with that I can check if the _update_tile_deferred is done with all the tiles by calling is_terrain_changeset_ready and just passing the whole Dictionary.

If that returns true, I can apply it with apply_terrain_changeset and save the Dictionary for later use, which comes in really handy!

I do the following now:
- Start a thread to calculate the island size and which tiles to place
- set the tiles via better_terrain.set_cells on the main thread
- after setting the layer, I call create_terrain_changeset and append it to an array
- in _process I check if the changeset is ready
- when the changeset is ready, I call apply_terrain_changeset

Which works great so far! EXCEPT one error I get randomly.
Every third or fourth time I get :
![image](https://github.com/Portponky/better-terrain/assets/11736148/acc3352c-846c-498e-9210-c5e5a91c11ba)

in _update_tile_deferred types: Array doesn't contain my coord, so it runs into an error.
I can prevent that by changing the line to:
`var type
	if types.has(coord):
		type = types[coord]
	else:
		return null`

Am just wondering why this is happening?

**Comment by Portponky** [comment]
You don't need to call `set_cells`, the changeset takes a dictionary of Vector2i coordinates, and integer terrain types. If you check the documentation (via F1 or in BetterTerrain.gd) it shows what to do. There's also an example in a video I made.

So you can skip the `set the tiles via better_terrain.set_cells on the main thread` step. Just create the dictionary, and create the changeset from the main thread.

**Comment by griessbreilp** [comment]
I solved my problem. And no, it wasn't your plugin, so the issue is resolved. >.<

For people running into a similar issue as I did, where you set the tiles (which works) but the update and fitting of the "ruletile" does not work; Maybe you are using floats.

In my generation method I read a JSON file where the settings are in. Like which tile has what noise_value, tile_id, layer_id, and all that.
JSON only knows "numbers". Neither int nor float. So Godot parses all of those as float.
_update_tile_tiles than checked if [5] has 5 and returned false. Because int 5 is not float 5. Which is absolutely correct.

I have a logic now in place, which makes sure that I actually hand over the tile_id as an int.
![image](https://github.com/Portponky/better-terrain/assets/11736148/88f1c6f3-4a51-4803-91a8-de0dbacfa678)

Thank you really much for your help!
